### PR TITLE
Preliminary Hough transform implementation

### PR DESCRIPTION
**Pull request by khilanravani**
We first extract some information from the image - width, height etc. , convert to luma and compute gradients. Then, for each pixel (x,y), we compute rho = xcos (theta) + ysin (theta). After a few more transformations and finding the bins with highest values, we generate the resulting image.

### CODE REVIEW

**Comment by lehins** [comment]
Good job @khilanravani, this looks like good progress. Besides the comments on the code itself here is one thing I'd like to see, namely Hough transform in action. Use a sample image and detect some lines in it, for instance, and then add the resulting image to the repo with example code how you generated it. For example you can use the `yield.jpg` image, that I think would work great for line detection, similar to how it's done in the documentation of this function: https://www.stackage.org/haddock/lts-10.10/hip-1.5.3.0/Graphics-Image-Processing-Binary.html#v:thresholdWith

**Comment by lehins** [comment]
> I didn't exactly understand if you want me to write a
separate function for detecting lines or you are asking me to use this
hough transform and then apply it to detect lines in an image.

What I mean is adding a good example that can be used in the documentation of the function, that demonstrates not only how to use the function but also the result produced by it. There is no need to add more sample images to the repo, that's why I suggested using `yield.jpg` file for the mentioned example.

Also what would be nice is to see Hough transform in action. Simply looking at an image produced by `hough` function isn't particularly interesting, would you agree? But if there is an example of actually using the result of Hough transform that would be awesome, eg. drawing detected lines on the source image would be cool. For example page 31 of this slides: https://www.uio.no/studier/emner/matnat/ifi/INF4300/h09/undervisningsmateriale/hough09.pdf
At the same time it might be a little bit of a complicated task for now and we can defer till later.

**Comment by khilanravani** [comment]
Sir,

> Also what would be nice is to see Hough transform in action. Simply
> looking at an image produced by hough function isn't particularly
> interesting, would you agree? But if there is an example of actually using
> the result of Hough transform that would be awesome, eg. drawing detected
> lines on the source image would be cool. For example page 31 of this
> slides: https://www.uio.no/studier/emner/matnat/ifi/INF4300/h09/
> undervisningsmateriale/hough09.pdf. At the same time it might be a little
> bit of a complicated task for now and we can defer till later.
>
Yes sir, I totally agree with you. It can be a good thing for people to
actually see the applications of Hough transform. Yes sir, exactly. It
might be a bit difficult to implement it right now. I'll try to spend some
time for doing it but if you I agree, it will be better if we postpone it
to a little later because actually, I'm running a bit behind my decided
timeline and 1st evaluations are coming pretty soon 😅. Although, I assure
you that I'll definitely work it out sooner or later. Kindly tell me what
you think of this.

Thank you,
Sincerely,
Khilan.

**Comment by lehins** [comment]
> I'll try to spend some time for doing it but if you I agree, it will be better if we postpone it

Yeah, that's fine. Go ahead and try resolving the rest of the comments I made and than I'll I'll guide you on further improvements that can be made.

**Comment by khilanravani** [comment]
Sure sir! Thank you!

**Comment by alpmestan** [comment]
@lehins Did you forget to send all your review comments?

File: src/Graphics/Image/Processing/Hough.hs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+import Data.Array.ST (newArray, writeArray, readArray, runSTArray)
+
+import Prelude as P hiding (subtract)
+import Graphics.Image
+import Graphics.Image.Interface as I
+import Graphics.Image.Types as IP
+
+-- | Some helper functions :
+-- | toImageY : Converts an image to Luma Image
+toImageY :: (ToY cs e, IP.Array arr cs e, IP.Array arr Y Double) =>
```
[the remark this answers is not in the document]
Comment (khilanravani) [reply src/Graphics/Image/Processing/Hough.hs]: Sorry sir. Will correct that.

File: src/Graphics/Image/Processing/Hough.hs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+-- | 'hough' computes the Linear Hough Transform and maps each point in the target image, ​ (ρ, θ) ​ 
+-- to the average color of the pixels on  the corresponding line of the source image ​(x,y) ​- space,
+-- where the line corresponds to points of the form ​(xcosθ + ysinθ = ρ(rho)). 
+--
+-- The idea is that where there is a straight line in the original image, it corresponds to a 
+-- bright (or dark, depending on the color of the background field) spot; by applying a suitable 
+-- filter to the results of the transform, it is possible to extract the locations of the lines in the original image.
+--
+-- <<images/frog_rbg.jpg>>
```
[the remark this answers is not in the document]
Comment (khilanravani) [reply src/Graphics/Image/Processing/Hough.hs]: Yes sir, I'll change it.

File: src/Graphics/Image/Processing/Hough.hs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+         PixelY y' = I.index luma (x, min (y+1) heightMax)
+     in (orig - x', orig - y')
+
+   slopeMap :: [ ((Int, Int), (Double, Double)) ]
+   slopeMap = [ ((x, y), slope x y) | x <- [0 .. widthMax], y <- [0 .. heightMax] ]
+
+   distMax :: Double -- Compute Maximum distance
+   distMax = (sqrt . fromIntegral $ (heightMax + 1) ^ (2 :: Int) + (widthMax + 1) ^ (2 :: Int)) / 2 
+
+   accBin = runSTArray $   -- Core part of Algo begins here. Working in a safe way with a mutable array.
```
[the remark this answers is not in the document]
Comment (khilanravani) [reply src/Graphics/Image/Processing/Hough.hs]: Ok sir.

File: src/Graphics/Image/Processing/Hough.hs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+   maxAcc = F.maximum accBin  
+   hTransform (x, y) =
+        let l = 255 - truncate ((accBin ! (x, y)) / maxAcc * 255) -- pixel generating function
+        in PixelRGB l l l
+
+   hImage :: Image arr RGB Word8
+   hImage = makeImage (thetaSz, distSz) hTransform
+
+
+test :: IO ()
```
[the remark this answers is not in the document]
Comment (khilanravani) [reply src/Graphics/Image/Processing/Hough.hs]: Sorry sir, that was for my testing. Forgot to remove it. Will do.

File: src/Graphics/Image/Processing/Hough.hs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  -> Image arr RGB Double
+hough image thetaSz distSz = I.map (fmap toDouble) hImage
+ where
+   widthMax, xCtr, heightMax, yCtr :: Int
+   widthMax = ((rows image) - 1)
+   xCtr = (widthMax `div` 2)
+   heightMax = ((cols image) - 1)
+   yCtr = (heightMax `div` 2)
+
+   luma :: Image arr Y Double
```
[the remark this answers is not in the document]
Comment (khilanravani) [reply src/Graphics/Image/Processing/Hough.hs]: Yes sir, actually earlier I used RGBA images but then switched to rgb. I'll change the input type.

File: src/Graphics/Image/Processing/Hough.hs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+-- >>>     houghImage = hough frog thetaSz distSz
+-- >>> writeImage "test.png" houghImage
+--
+hough
+  :: forall arr a.
+     ( IP.Array arr RGB a, IP.Array arr RGB Word8
+     , MArray arr Y Double, IP.Array arr Y Double
+     , IP.Array arr RGB Double
+     )
+  => Image arr RGB a
```
[the remark this answers is not in the document]
Comment (khilanravani) [reply src/Graphics/Image/Processing/Hough.hs]: Sir, as per my understanding, Hough transform is most commonly used in combination with canny, sobel or sometimes with de-noising filters. In that case, the input image is generally pre-processed and converted to binary or grayscale (more preferred). The basic concept of Hough doesn't limit its usage to binary images only although it depends on what exactly we use it for. I checked some other implementations and found that they also use grayscale images in most cases. So, I guess taking rgba as input, converting to luma and applying the algorithm shouldn't be a problem. Please correct me if I'm misunderstanding some point here.

File: src/Graphics/Image/Processing/Hough.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (lehins) [reply src/Graphics/Image/Processing/Hough.hs]: What I meant by the comment above is that if `hough` works with luma images, then it should accept luma images not `RGB` or `RGBA`, it needs `Y`. So you don't need to use `toImageY`, the user of the `hough` function will. Here is what a proposed type signature might look like:

```
hough
  :: ( MArray arr Y Double, IP.Array arr Y Double )
  => Image arr Y Double
  -> Int
  -> Int
  -> Image arr Y Double
```

Unless I am mistaken, you don't need `forall arr a`, nor do you need either one of `RankNTypes` or `ScopedTypeVariables` extension up top.

[the remark this answers is not in the document]
Comment (lehins) [reply src/Graphics/Image/Processing/Hough.hs]: I was mistaken, I see now, you do need `forall arr` and `ScopedTypeVariables`, but not the `RankNTypes`

So the signature will be something along those lines:

```
hough
  :: forall arr . ( MArray arr Y Double, IP.Array arr Y Double, IP.Array arr Y Word8)
  => Image arr Y Double
  -> Int
  -> Int
  -> Image arr Y Word8
```

File: src/Graphics/Image/Processing/Hough.hs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+-- | 'hough' computes the Linear Hough Transform and maps each point in the target image, ​ (ρ, θ) ​ 
+-- to the average color of the pixels on  the corresponding line of the source image ​(x,y) ​- space,
+-- where the line corresponds to points of the form ​(xcosθ + ysinθ = ρ(rho)). 
+--
+-- The idea is that where there is a straight line in the original image, it corresponds to a 
+-- bright (or dark, depending on the color of the background field) spot; by applying a suitable 
+-- filter to the results of the transform, it is possible to extract the locations of the lines in the original image.
+--
+-- <<images/yield.jpg>>
+--
```
Comment (lehins) [inline src/Graphics/Image/Processing/Hough.hs]: Could you please include the result image of applying the hough transform to `yield.jpg` here.
This way a user will know what sort of result to expect.

Comment (khilanravani) [reply src/Graphics/Image/Processing/Hough.hs]: Sure sir, will do.

File: src/Graphics/Image/Processing/Hough.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (khilanravani) [reply src/Graphics/Image/Processing/Hough.hs]: Ok sir, I'll change it accordingly.

File: src/Graphics/Image/Processing/Hough.hs (hunk 7 above)
Comment (khilanravani) [reply src/Graphics/Image/Processing/Hough.hs]: Sure sir, will do so.

File: src/Graphics/Image/Processing/Hough.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (khilanravani) [reply src/Graphics/Image/Processing/Hough.hs]: Ok sir, I'll change that.

**Pull request merged**